exercicios-yuri/exercicio-api/main.go

```diff
@@ -0,0 +1,250 @@
+	"errors"
+	"log"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gin-gonic/gin"
+)
+
+func main() {
+
+	u := newItemUsecase()
+	h := newHandler(u)
+
+	router := gin.Default()
+
+	router.GET("/", h.home)
+	router.GET("/hello", h.hello)
+	router.POST("/bye", h.bye)
+
+	router.POST("/items", h.saveItem)
+	router.GET("/items", h.listItems)
+
+	router.GET("/items/:id", h.itemById)
+	router.PUT("/items", h.changeItem)
+	router.DELETE("/items/:id", h.deleteItem)
+
+	log.Println("Server started at http://localhost:8080/")
+
+	if err := router.Run(":8080"); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// /////////////////////////////////////////////////////////////////////////////
+// Global error
+// /////////////////////////////////////////////////////////////////////////////
+var errNotFound = errors.New("not found")
+
+// ////////////////////////////////////////////////////////////////////////////
+// Handler
+// ////////////////////////////////////////////////////////////////////////////
+type handler struct {
+	usecase *itemUsecase
+}
+
+// Handler constructor
+func newHandler(u *itemUsecase) *handler {
+	return &handler{
+		usecase: u,
+	}
+}
+
+func (h *handler) home(c *gin.Context) {
+	c.String(http.StatusOK, "Welcome to the home page!")
+}
+
+func (h *handler) hello(c *gin.Context) {
+	c.String(http.StatusOK, "Hello, world!")
+}
+
+func (h *handler) bye(c *gin.Context) {
+	var msg map[string]string
+	if err := c.BindJSON(&msg); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to bind JSON"})
+		return
+	}
+	message, exists := msg["message"]
+	if !exists {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Message field is missing"})
+		return
+	}
+	c.String(http.StatusOK, "Received POST request with message: %s", message)
+}
+
+func (h *handler) saveItem(c *gin.Context) {
+	var it item
+
+	if err := c.BindJSON(&it); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	if err := h.usecase.saveItem(it); err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, "item saved successfully")
+}
+
+func (h *handler) listItems(c *gin.Context) {
+	its, err := h.usecase.listItems()
+	if err != nil {
+		if err == errNotFound {
+			c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
+		} else {
+			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		}
+		return
+	}
+
+	c.JSON(http.StatusOK, its)
+}
+
+func (h *handler) itemById(c *gin.Context) {
+	idStr := c.Param("id")
+	id, err := strconv.Atoi(idStr)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
+		return
+	}
+
+	item, errUsecase := h.usecase.itemById(id)
+
+	if errUsecase != nil {
+		if errUsecase == errNotFound {
+			c.JSON(http.StatusNotFound, gin.H{"error": errUsecase.Error()})
+		} else {
+			c.JSON(http.StatusInternalServerError, gin.H{"error": errUsecase.Error()})
+		}
+		return
+	}
+	c.JSON(http.StatusOK, item)
```

Comment:
Por lo general primero valdamos los errores, si estos no son nil cortamos el proceso, si es nil continua el proceso.

```suggestion
	    if errUsecase == nil {
        c.JSON(http.StatusOK, item)
        return
    }

    if errUsecase == errNotFound {
        c.JSON(http.StatusNotFound, gin.H{"error": errUsecase.Error()})
    } else {
        c.JSON(http.StatusInternalServerError, gin.H{"error": errUsecase.Error()})
    }
```